alf/metrics/metrics.py

```diff
@@ -0,0 +1,251 @@
+        self._return_accumulator[:] = torch.where(
+            time_step.is_first(), torch.zeros_like(self._return_accumulator),
+            self._return_accumulator)
+
+        # Update accumulator with received rewards.
+        # Ignores first step whose reward comes from the boundary transition
+        # of the last step from the previous episode.
+        self._return_accumulator.add_(
+            torch.where(time_step.is_first(),
+                        torch.zeros_like(self._return_accumulator),
+                        time_step.reward))
+
+        # Add final returns to buffer.
+        last_episode_indices = torch.squeeze(
+            *torch.where(time_step.is_last())).type(torch.int64)
+        for indx in last_episode_indices:
+            self._buffer.append(
+                self._return_accumulator[indx].clone().detach())
+
+        return time_step
+
+    def result(self):
+        return self._buffer.mean()
+
+    def reset(self):
+        self._buffer.clear()
+        self._return_accumulator.fill_(0)
+
+
+class AverageEpisodeLengthMetric(metric.StepMetric):
+    """Metric to compute the average episode length."""
+
+    def __init__(self,
+                 name='AverageEpisodeLength',
+                 prefix='Metrics',
+                 dtype=torch.float32,
+                 batch_size=1,
+                 buffer_size=10):
+        super(AverageEpisodeLengthMetric, self).__init__(
+            name=name, prefix=prefix)
+        # TODO: use tensor deque
+        self._buffer = THDeque(max_len=buffer_size, dtype=dtype)
+        self.dtype = dtype
+        self._length_accumulator = torch.nn.Parameter(
+            torch.zeros(batch_size, dtype=dtype), requires_grad=False)
+
+    def call(self, time_step):
+        """Accumulates the number of episode steps according to batched time_step.
+        It would increase for all non time_step.is_first().  The first time_step
+        is the boundary step and needs to be ignored, different from tf_agents.
+
+        Args:
+            time_step (alf.data_structures.TimeStep): batched tensor
+        Returns:
+            The arguments, for easy chaining.
+        """
+        # Each non-boundary time_step (mid or last) represents a step.
+        is_first = time_step.is_first()
+        non_boundary = torch.where(
+            torch.logical_not(is_first),
+            torch.ones_like(self._length_accumulator),
+            torch.zeros_like(self._length_accumulator))
+        self._length_accumulator.add_(non_boundary)
+
+        # Add lengths to buffer when we hit end of episode
+        is_last = time_step.is_last()
+        last_indices = torch.squeeze(*torch.where(is_last)).type(torch.int64)
+        for indx in last_indices:
+            self._buffer.append(
+                self._length_accumulator[indx].clone().detach())
+
+        # Clear length accumulator at the end of episodes.
+        self._length_accumulator[:] = torch.where(
```

Comment:
It's better to use self._length_accumulator.scatter_() to do the update. Similar to the tf_agent implementation

Reply:
Done